packages/@aws-cdk/core/lib/arn.ts

```diff
@@ -76,11 +76,20 @@ export class Arn {
    * the 'scope' is attached to. If all ARN pieces are supplied, the supplied scope
    * can be 'undefined'.
    */
-  public static format(components: ArnComponents, stack: Stack): string {
-    const partition = components.partition !== undefined ? components.partition : stack.partition;
-    const region = components.region !== undefined ? components.region : stack.region;
-    const account = components.account !== undefined ? components.account : stack.account;
-    const sep = components.sep !== undefined ? components.sep : '/';
+  public static format(components: ArnComponents, stack?: Stack): string {
+    const requiredComponents: Array<keyof ArnComponents> = ['partition', 'region', 'account'];
+    if (stack === undefined) {
+      requiredComponents.forEach(c => {
+        if (components[c] === undefined) {
+          throw new Error(`Stack must be defined if "${c}" component is undefined`);
+        }
+      });
+    }
+
+    const partition = components.partition ?? stack!.partition;
```

Comment:
`!` should not be needed here and generally discouraged in our code base

Reply:
Unfortunately it is. I think as of TypeScript 3.9 we can remove it due to better `Error` tracking, but until then it's needed.

Reply:
Then I rather we simplify the code to something less smart:

```ts
const componentRequiredError = c => new Error(`"stack" must be defined since "${c}" is not defined`);
if (components.partition === undefined) { throw componentRequiredError('partition'); }
if (components.region === undefined)    { throw componentRequiredError('region');    }
if (components.account === undefined)   { throw componentRequiredError('account');   }
```

I know, less sexy...

Reply:
?